Pre RFC: Macros 1.1

alexcrichton opened the pull request:
[Explanation here](https://github.com/alexcrichton/rfcs/pull/4#issuecomment-233031985)

[Rendered](https://github.com/alexcrichton/rfcs/blob/plugins-1.1/text/0000-macros-1.1.md)

alexcrichton committed:
RFC: Plugins 1.1

Stabilize a very small sliver of today's plugin system in the compiler, just
enough to get basic features like custom derive working. Ensure that the
features stabilized here will not pose a maintenance burden on the compiler but
also don't try to stabilize enough features for the "perfect macro system" at
the same time. Overall, this should be considered an incremental step towards an
official "plugins 2.0".

alexcrichton committed:
s/plugin/macro/g

alexcrichton committed:
Specifically allow custom attributes

alexcrichton committed:
Add a section about spans

alexcrichton commented:
Hey everyone cc'd here! I've discussed this briefly with some of you, but I wanted to make a spot where we can all talk about it (and cc more plugin authors as well). This is a pre-RFC that we've been thinking about recently which should help pave the way for stabilizing procedural macros as they are today in the very near future. The main goal here is to get large projects like serde/diesel in their "best form" working on stable.

I'll leave most of the explanation to the RFC itself, though! I intend to send this PR to rust-lang/rfcs eventually but I want to make sure that everyone's on board with how this could work today, wouldn't want to leave out anything important by accident!

Here's a [rendered version](https://github.com/alexcrichton/rfcs/blob/plugins-1.1/text/0000-macros-1.1.md)

Here's [the PR](https://github.com/alexcrichton/rfcs/pull/4)

Feel free to leave comments here and/or on commits!

cc list -- @erickt, @sgrif, @dtolnay, @oli-obk, @nrc, @aturon, @nikomatsakis, @cgswords, @brson, @jonathandturner, @pcwalton, @steveklabnik 

Note that this is still a somewhat early draft, so I'd appreciate it if it weren't shared _everywhere_ just yet, but cc'ing more authors should be ok if I forgot someone!

alexcrichton commented:
@nrc so to respond to a piece of feedback you had in email:

> I'd like to see this restricted to derive, rather than proc macros in general. That would address a large deal of the pain (all of Serde, much of diesel) and keep the temporary solution small (it also side-steps the question of import/export/modularisation). I.e., we only support rustc_macro_derive

I talked with @sgrif on IRC and got [this feedback](https://gist.github.com/sgrif/11c49e8ba6c28ec18c8d00b4c88f5d64). After some more discussion, it sounded like the macros like `infer_schema!` were used by almost everyone using diesel, so important to stabilize for the _best_ experience.

Could you elaborate a little more on how we can maybe future-proof these kinds of macros? They're obviously pretty un-ideal today, but perhaps there are some more restrictions/safeguards we should put in place to hinder them? For example should they only be expand-able to items, and they can only be invoked directly inside of modules? (not sure if that'd actually help)

alexcrichton commented:
cc @eddyb

eddyb commented:
I am okay with stabilizing everything proposed here, _except_ `#![rustc_macro_crate]` - could we not load the macro crate as a plugin from just an `extern crate`, with the existing import attributes for macros?

eddyb commented on text/0000-macros-1.1.md:
```diff
@@ -0,0 +1,443 @@
+# ...
+
+[lib]
+rustc-macro = true
```
Does Cargo not expose crate-type itself? Was a Cargo change necessary to make cdylibs work, for example?

alexcrichton replied:
Cargo would indeed accept 

``` toml
crate-type = ["rustc-macro"]
```

although it wouldn't work quite right because Cargo needs extra knowledge that it's being used as a macro crate, so for cross compilations it needs to compile it and all of its dependencies for the host architecture rather than the target. 

With that extra piece of information I figured it'd be better to have just one attribute in the TOML rather than two. (also avoids Cargo trying to interpret crate types to figure out of it's being used as a plugin or a dependency)

eddyb replied:
Alright, that makes sense.

eddyb commented on text/0000-plugins-1.1.md:
```diff
@@ -0,0 +1,486 @@
+- Start Date: 2016-07-14
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
```
Should this copy of the RFC (with "plugins" in the file name) still be part of the diff?

alexcrichton replied:
Oops, nope!

alexcrichton committed:
Remove 0000-plugins-1.1.md

alexcrichton commented:
@eddyb 

> could we not load the macro crate as a plugin from just an extern crate, with the existing import attributes for macros?

I thought about this, yeah, but my preference is to avoid it because `extern crate foo` looks like I can import things from foo (e.g. `foo::bar::baz`), but with plugin crates you can't do that because they're not even compiled for the right architecture. I kinda assume that macros 2.0 will solve this and use `extern crate` instead, but I'm not entirely certain what the story there will be.

Does that make sense? I should certainly add it as an alternative at least!

alexcrichton committed:
Small updates:

* Add MacroContext
* alternative syntax for defining macros
* unresolved question of how close can we get to macros 2.0

nrc commented:
@erickt @alexcrichton I'd be keen to avoid any span info here to allow us to evolve spans without any risk of breaking change. It means a smaller surface error for problems and more motivation to abandon these macros in the future.

nrc commented:
@alexcrichton re non-derive or extending derive:

I think we should aim for a maximally minimal RFC here. Less to argue about, quicker road to stabilisation. We should be clear that we are trying to provide an 80/20 kind of solution here, not something that fills all use cases. We could consider further RFCs to expand macros 1.1 (and should make that clear in this RFC). In particular sgriff's suggestion to extend derive with arguments seems like something we probably want at some point, but could be a big design space, and I'd want to think about any plan there in terms of short- and long-term plans.

I really think we shouldn't stabilise any macros other than derive. I don't think giving the best experience to users of a single library should motivate changes like this to the language. I would rather weaken our stability checking or something rather than stabilise all macros like this - even though its a designed to be low maintenance, it is still cruft we'll be stuck with for years.

I don't think there is much we can do in the design space to shrink the impact either. Item vs expression doesn't really make any difference.

I am OK with namespacing in derive, that seems like a no-brainer for something we'd want.

nrc commented:
@alexcrichton @eddyb 

The plan for macros 2.0 is to just use `extern crate foo` but `foo::...` would not work because it is a macro crate, possibly loosened in the future. I don't feel that is too bad - it does change the intuition of `extern crate` a little bit, but I feel like expecting users to know a little about the crate type is Ok.

So I would also like to use just `extern crate` here, since we will use that approach in macros 2.0. We will get a chance to see how much confusion that causes - we are aiming for quick stabilisation, not insta-stable, right?

nrc commented on text/0000-macros-1.1.md:
```diff
@@ -0,0 +1,473 @@
+The good news, however, is that this class of projects which require nightly
+Rust almost all require nightly for the reason of procedural macros. Even
+better, the full functionality of procedural macros is rarely needed, only
+custom derive! The purpose of this RFC, as a result, is to provide these crates
```
Might be nice to note why derive is a nice sub-case of macros - hygiene is not really an issue, modularisation is much easier.

alexcrichton replied:
Excellent points!

nrc commented on text/0000-macros-1.1.md:
```diff
@@ -0,0 +1,473 @@
+}
+
+#[derive(Debug)]
+pub struct FromSourceError {
```
I would call this `LexError`, but minor

nrc commented on text/0000-macros-1.1.md:
```diff
@@ -0,0 +1,473 @@
+    // ...
+}
+
+impl TokenStream {
```
I think these need to be on `TokenSlice` and we need that type + `Deref` for TokenStream to TokenSlice (think str vs String).

nrc replied:
Also, I expect that from_source will need to take the MacroContext as an arg

alexcrichton replied:
After the discussion with @eddyb on IRC, I think the conclusion was that we may not want to add `TokenSlice` just yet but I'm totally fine adding cx as an arg to `from_source`. From an API point of view we should still be able to add `TokenSlice` at any time backwards compatibly as well I believe.

nrc commented on text/0000-macros-1.1.md:
```diff
@@ -0,0 +1,473 @@
+#![crate_type = "rustc-macro"]
+#![crate_name = "double"]
+
+extern crate macro;
```
I think we need a better name for libmacro since `macro` is a keyword, although I guess we could make it contextual (I only want to use it for `macro!`). It is currently reserved.

alexcrichton replied:
Eh, "rustc_macro" may be good enough

nrc commented on text/0000-macros-1.1.md:
```diff
@@ -0,0 +1,473 @@
+}
+
+#[rustc_macro_derive(Double)]
+pub fn double(_cx: &mut MacroContext, input: TokenStream) -> TokenStream {
```
I wonder if we should abstract the `&mut MacroContext` into a type (`MacroContextRef` or something) in case we need to change to `Rc` or something down the line?

alexcrichton replied:
I don't mind either way, just following the blog post here!

nrc commented on text/0000-macros-1.1.md:
```diff
@@ -0,0 +1,473 @@
+  and is significantly harder to write. The marginal benefit of being slightly
+  more flexible about how it's run likely isn't worth it.
+
+* The syntax for defining a macro will likely be different in the macros 2.0
```
I'm not sure I understand this point - I think macros 2.0 is going to use an attribute on a function.

alexcrichton replied:
Oh I think @cgswords was under the impression that procedural macros 2.0 would be something like `pub macro foo(cx: &mut Context, TokenStream) -> TokenStream`, but we may be misinterpreting!

alexcrichton commented:
@nrc 

> I'd be keen to avoid any span info here to allow us to evolve spans without any risk of breaking change

Oh yeah definitely! We'll have to do _something_ in `from_source` but I think it should be fine to just attach it to the "current span" in the context passed in. Sounds like we can eventually stabilize more APIs here to provide all the necessary functionality.

> re non-derive or extending derive:

We talked on IRC about this, the conclusion being that I'll see if we can finagle it out for now. I'm not 100% convince we can, but I'm not 100% convinced we can't either.

> I am OK with namespacing in derive, that seems like a no-brainer for something we'd want.

Oh I think I missed this, could you elaborate on this? Right now this RFC doesn't propose any namespacing, did you have some in mind though?

> So I would also like to use just extern crate here, since we will use that approach in macros 2.0. 

My hesitation here is:
- The import issue I mentioned before, although you're right that it may not actually cause confusion.
- The compiler doesn't know what to look for. If it finds both a host and target version of the crate, which does it choose? How does it know which you intended? Using `#![rustc_macro_crate]` clearly indicates which, but it is indeed not super ergonomic.

Thoughts on that second point?

> We will get a chance to see how much confusion that causes - we are aiming for quick stabilisation, not insta-stable, right?

Definitely, yeah.

alexcrichton committed:
More updates:

* librustc_macro, not libmacro
* custom attributes are allowed by having modifiers everywhere, not decorators

alexcrichton committed:
Alternative of TLS

alexcrichton commented:
@sgrif wanna ping me on IRC tmw, got a few things I wanna chat about

alexcrichton committed:
Alternative of `extern crate`

alexcrichton committed:
one more extern crate rustc_macro

alexcrichton committed:
more s/macro/rustc_macro/

alexcrichton committed:
spacing

eddyb commented on text/0000-macros-1.1.md:
```diff
@@ -0,0 +1,494 @@
+  generate an error if you import paths from it, so it doesn't behave like other
+  `extern crate` statements, and the compiler also doesn't know whether to look
+  for a target or host crate when it encounters such a definition. If both are
+  found it doesn't know which is the right to choose as well.
```
I think this is the only real downside. Could `Cargo` handle this correctly by default?

eddyb commented:
There is a bit of a problem with the string-based expansion: there might be a lossy conversion from interpolated tokens, e.g.:

``` rust
macro_rules! foo {
    ($x:expr) => {
        #[derive(Serialize)]
        enum Foo { Bar = $x, Baz = $x * 2 }
    }
}
foo!(1 + 1);
```

In the current implementation, this might output `Baz = 1 + 1 * 2`, i.e. `Baz = 3`, even though `rustc` would parse it as `Baz = (1 + 1) * 2` i.e. `Baz = 4`. If `serde` actually uses that value, it wouldn't match.

In cases such as expressions or types, we can parenthesize them, and some of the other situations don't result in ambiguities, but this is something we have to be wary of.

**EDIT**: You can observe [the current implementation for source stringification](https://play.rust-lang.org/?gist=cb6a8ca5707f13d0bfc6f5fdfbdb7a59&version=nightly&backtrace=1&run=1).

alexcrichton committed:
s/MacroContext/Context/

alexcrichton commented:
@eddyb good point, I'll be sure to add a note.

sfackler commented:
WRT namespaceing, I'd love to initially just namespace by the plugin crate name e.g. `#[derive(serde_macros::Deserialize)]`. We can require that all `#[rustc_macro_derive]`s are defined in the crate root to future proof things. Dunno if we'd need to add some ability to `use` derives, but I personally don't see a need until we can do it "for real".

alexcrichton committed:
Remove procedural macros

alexcrichton commented:
@sfackler I'm currently trying to keep this RFC to have no dependencies on other RFCs, and I don't think that's possible yet?

It's also unclear whether we'd also want things like `#[derive(::serde_macros::Deserialize)]` vs without the leading `::`, so maybe there's enough open questions to just have a global namespace?

I'm under the impression that we'll eventually have two forms of defining custom derive modes if this all stabilizes. One is all this, and another is a properly scoped and imported version (tbd)

alexcrichton committed:
Another drawback

alexcrichton committed:
Add alternative of namespaces

alexcrichton commented:
Ok, I've added a number of drawbacks and alternatives and also removed `#[rustc_macro_define]`, that is only custom derive modes are allowed now.

I'll probably post later today pending any objections coming up, thanks for the discussions again everyone!

eddyb commented on text/0000-macros-1.1.md:
```diff
@@ -492,3 +520,9 @@ pub struct Foo {
+
+* Eventually macro crates will want to be loaded both at compile time and
+  runtime, and this means that Cargo will need to understand to compile these
+  crates twice, once as `rustc-macro` and once as an rlib. Does Cargo have
```
Couldn't `rustc` do that? I guess there would be complications related to host/target separation. Not sure if @nrc had any solution in mind.

alexcrichton replied:
Perhaps, yeah, but the compiler would have to run itself twice basically anyway I think because:
- What's actually compiled changes
- LLVM can only generate one target at a time
- The dependencies for each target may be different

etc. Some of this definitely seems like Cargo's domain?

alexcrichton commented:
Ok, I'm going to close this now, and an RFC will soon be open on rust-lang/rfcs, let's continue discussion there.

nrc commented:
> >    I am OK with namespacing in derive, that seems like a no-brainer for something we'd want.
> 
> Oh I think I missed this, could you elaborate on this? Right now this RFC doesn't propose any namespacing, did you have some in mind though?

This was in response to a comment in the gist from sgriff. It seems something we want, if we want to derive for a non prelude trait, then you need to specify the path to the trait

nrc commented:
> ```
> The compiler doesn't know what to look for. If it finds both a host and target version of the crate, which does it choose? How does it know which you intended? Using #![rustc_macro_crate] clearly indicates which, but it is indeed not super ergonomic.
> ```

The compiler should open the crate and if it is a macro crate then it must be for the host and if it is non-macro then it should be a target crate. There should never be different stuff in the crate, so there shouldn't ever be ambiguity.

For macros 2.0, I proposed designating whether a crate was a macro crate or not with `cfg`, but for a quick solution if the compiler finds any macro attribute in the crate then it adds a macro bit to the metadata.
